R/verb-fill.R

```diff
@@ -0,0 +1,195 @@
+#'   1,      "Tom",    "Observer",   NA,                 4,
+#'   2,     "Mike",    "Observer",   NA,                NA,
+#'   2,  "Rachael",    "Observer",   NA,                 6,
+#'   2,  "Sydekea", "Scorekeeper",   14,                NA,
+#'   2, "Gabriela",    "Observer",   NA,                NA,
+#'   3,  "Derrick",    "Observer",   NA,                NA,
+#'   3,     "Kara", "Scorekeeper",    9,                 10,
+#'   3,    "Emily",    "Observer",   NA,                NA,
+#'   3, "Danielle",    "Observer",   NA,                NA
+#' )
+#' squirrels$id <- 1:12
+#'
+#' if (require("tidyr", quietly = TRUE)) {
+#'   tbl_memdb(squirrels) %>%
+#'     window_order(id) %>%
+#'     tidyr::fill(
+#'       n_squirrels,
+#'       n_squirrels2,
+#'     )
+#' }
+fill.tbl_lazy <- function(.data, ..., .direction = c("down", "up")) {
+  sim_data <- simulate_vars(.data)
+  cols_to_fill <- syms(names(tidyselect::eval_select(expr(c(...)), sim_data)))
+  order_by_cols <- op_sort(.data)
+  .direction <- arg_match0(.direction, c("down", "up"))
+
+  if (is_empty(order_by_cols)) {
+    abort(
+      c(
+        x = "`.data` does not have explicit order.",
+        i = "Please use arrange() or window_order() to make determinstic."
+      )
+    )
+  }
+
+  if (.direction == "up") {
+    # `desc()` cannot be used here because one gets invalid SQL if a variable
+    # is already wrapped in `desc()`
+    # i.e. `desc(desc(x))` produces `x DESC DESC`
+    order_by_cols <- purrr::map(order_by_cols, ~ quo(-!!.x))
+  }
+
+  dbplyr_fill0(
+    .con = remote_con(.data),
+    .data = .data,
+    cols_to_fill = cols_to_fill,
+    order_by_cols = order_by_cols,
+    .direction = .direction
+  )
+}
+
+dbplyr_fill0 <- function(.con, .data, cols_to_fill, order_by_cols, .direction) {
+  UseMethod("dbplyr_fill0")
+}
+
+#' @export
+dbplyr_fill0.DBIConnection <- function(.con,
+                                          .data,
+                                          cols_to_fill,
+                                          order_by_cols,
+                                          .direction) {
+  # strategy:
+  # 1. construct a window
+  # * from the first row to the current row
+  # * ordered by `order_by_cols`
+  # * partitioned by groups if any
+  # 2. in this window use the last value ignoring nulls; in SQL this is (usually)
+  #    `LAST_VALUE(<column> IGNORE NULLS)`
+  #
+  # Source: https://dzone.com/articles/how-to-fill-sparse-data-with-the-previous-non-empt
+  grps <- op_grps(.data)
+
+  fill_sql <- purrr::map(
+    cols_to_fill,
+    ~ win_over(
+      last_value_sql(.con, .x),
+      partition = if (!is_empty(grps)) escape(ident(op_grps(.data)), con = .con),
+      order = translate_sql(!!!order_by_cols, con = .con),
+      con = .con
+    )
+  ) %>%
+    set_names(as.character(cols_to_fill))
+
+  .data %>%
+    transmute(
+      !!!syms(colnames(.data)),
+      !!!fill_sql
+    )
+}
+
+#' @export
+dbplyr_fill0.SQLiteConnection <- function(.con,
```

Comment:
Just mention somewhere that this is a fall back for databases that don't have IGNORE NULLs support?